parser/src/main/scala/org/bykn/bosatsu/parser/Parser.scala

```diff
@@ -0,0 +1,938 @@
+    // we use defer to avoid enumerating this function unless needed
+    defer1(charIn1(Impl.allChars.filter(fn)))
+
+  def void[A](pa: Parser[A]): Parser[Unit] =
+    pa match {
+      case v@Impl.Void(_) => v
+      case notVoid => Impl.Void(Impl.unmap(pa))
+    }
+
+  def void1[A](pa: Parser1[A]): Parser1[Unit] =
+    pa match {
+      case v@Impl.Void1(_) => v
+      case notVoid => Impl.Void1(Impl.unmap1(pa))
+    }
+
+  def string[A](pa: Parser[A]): Parser[String] =
+    pa match {
+      case str@Impl.StringP(_) => str
+      case notVoid => Impl.StringP(Impl.unmap(pa))
+    }
+
+  def string1[A](pa: Parser1[A]): Parser1[String] =
+    pa match {
+      case str@Impl.StringP1(_) => str
+      case notVoid => Impl.StringP1(Impl.unmap1(pa))
+    }
+
+  val index: Parser[Int] =
+    Impl.Index
+
+  // succeeds when we are at the start
+  val start: Parser[Unit] = Impl.StartParser
+
+  // succeeds when we are at the end
+  val end: Parser[Unit] = Impl.EndParser
+
+  /**
+   * If we fail, rewind the offset back so that
+   * we can try other branches. This tends
+   * to harm debuggability and ideally should be
+   * minimized
+   */
+  def backtrack[A](pa: Parser[A]): Parser[A] =
+    pa match {
+      case bt: Impl.Backtrack[A] => bt
+      case bt: Impl.Backtrack1[A] => bt
+      case nbt => Impl.Backtrack(nbt)
+    }
+
+  /**
+   * If we fail, rewind the offset back so that
+   * we can try other branches. This tends
+   * to harm debuggability and ideally should be
+   * minimized
+   */
+  def backtrack1[A](pa: Parser1[A]): Parser1[A] =
+    pa match {
+      case bt: Impl.Backtrack1[A] => bt
+      case nbt => Impl.Backtrack1(nbt)
+    }
+
+  implicit val catsInstancesParser1: FlatMap[Parser1] with Defer[Parser1] with MonoidK[Parser1]=
+    new FlatMap[Parser1] with Defer[Parser1] with MonoidK[Parser1] {
+      def empty[A] = Fail
+
+      def defer[A](pa: => Parser1[A]): Parser1[A] =
+        defer1(pa)
+
+      def map[A, B](fa: Parser1[A])(fn: A => B): Parser1[B] =
+        map1(fa)(fn)
+
+      def flatMap[A, B](fa: Parser1[A])(fn: A => Parser1[B]): Parser1[B] =
+        flatMap10(fa)(fn)
+
+      override def product[A, B](pa: Parser1[A], pb: Parser1[B]): Parser1[(A, B)] =
+        product10(pa, pb)
+
+      override def map2[A, B, C](pa: Parser1[A], pb: Parser1[B])(fn: (A, B) => C): Parser1[C] =
+        map(product(pa, pb)) { case (a, b) => fn(a, b) }
+
+      override def map2Eval[A, B, C](pa: Parser1[A], pb: Eval[Parser1[B]])(fn: (A, B) => C): Eval[Parser1[C]] =
+        Now(pb match {
+          case Now(pb) => map2(pa, pb)(fn)
+          case later => map2(pa, defer(later.value))(fn)
+        })
+
+      def tailRecM[A, B](init: A)(fn: A => Parser1[Either[A, B]]): Parser1[B] =
+        tailRecM1(init)(fn)
+
+      def combineK[A](pa: Parser1[A], pb: Parser1[A]): Parser1[A] =
+        Parser.oneOf1(pa :: pb :: Nil)
+
+      override def void[A](pa: Parser1[A]): Parser1[Unit] =
```

Comment:
We should override as and <* and *> to leverage void. So as(a) = void.map(_ => a)